### PR TITLE
steelseries: Add hid support to detect Bluetooth devices

### PR DESCRIPTION
**Pull request by gportay**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by gportay** [comment]
@hughsie: Could you tell what I am doing wrong with this?

```
sudo -E  build/src/fwupdtool get-devices  
Loading…                 [-                                      ]14:57:03:0652 FuUsbDevice          fu_usb_device_incorporate: assertion 'FU_IS_USB_DEVICE(donor)' failed
14:57:03:0652 libgusb              g_usb_device_get_vid: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 libgusb              g_usb_device_get_release: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 libgusb              g_usb_device_get_vid: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 libgusb              g_usb_device_get_pid: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 libgusb              g_usb_device_get_interfaces: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 libgusb              g_usb_device_release_interface: assertion 'G_USB_IS_DEVICE (device)' failed
14:57:03:0652 GLib                 g_propagate_error: assertion 'src != NULL' failed
**
GLib:ERROR:../glib/glib/gerror.c:1152:g_propagate_prefixed_error: assertion failed: (*dest != NULL)
Bail out! GLib:ERROR:../glib/glib/gerror.c:1152:g_propagate_prefixed_error: assertion failed: (*dest != NULL)
make: *** [Makefile:155: tool-get-devices] Aborted
```

**Comment by gportay** [comment]
I would like to detect the mouse connected via BT using the HID device; and send a Report command to get the version. I am not able to detect it. My knowlegde is very limited for now about HID devices.

File: plugins/steelseries/steelseries.quirk [hunk 1]
```diff
@@ -12,6 +12,12 @@ GType = FuSteelseriesSonic
 Icon = input-mouse
 
 # Aerox 3 Wireless
+[HIDRAW\VEN_0111]
```
Comment (hughsie) [inline plugins/steelseries/steelseries.quirk:15]: this seems over-broad; I think you want a DEV_ there, no? At the moment it's matching all HID devices with that vendor.

File: plugins/steelseries/fu-steelseries-hid.c [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		fu_common_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+
+	/* success */
+	return g_strndup((const gchar *)data, sizeof(data));
+}
+
+static gboolean
+fu_steelseries_hid_probe(FuDevice *device, GError **error)
+{
+	/* FuUsbDevice->probe */
```
Comment (hughsie) [inline plugins/steelseries/fu-steelseries-hid.c:77]: FuHidDevice in the comment pls

File: plugins/steelseries/fu-steelseries-hid.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	/* set the physical ID */
+	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "hid", error);
+}
+
+static gboolean
+fu_steelseries_hid_setup(FuDevice *device, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	/* FuUsbDevice->device*/
```
Comment (hughsie) [inline plugins/steelseries/fu-steelseries-hid.c:90]: FuHidDevice in the comment pls

File: plugins/steelseries/fu-steelseries-hid.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+	if (!FU_DEVICE_CLASS(fu_steelseries_hid_parent_class)->setup(device, error))
+		return FALSE;
+
+	g_debug("Version: '%s'\n", fu_steelseries_hid_version(device, &error_local));
```
Comment (hughsie) [inline plugins/steelseries/fu-steelseries-hid.c:94]: not using fu_device_set_version here?

**Comment by hughsie** [comment]
> Could you tell what I am doing wrong with this

Can you attach the whole log please; I think it need a bit more context.

File: plugins/steelseries/steelseries.quirk (hunk 1 above)
Comment (gportay) [reply plugins/steelseries/steelseries.quirk:15]: Sure, this is a draft; I have voluntary dropped for the testing.

File: plugins/steelseries/fu-steelseries-hid.c (hunk 2 above)
Comment (gportay) [reply plugins/steelseries/fu-steelseries-hid.c:77]: yes sure; lots of changes to figure out how to detect the HID device over BT.

File: plugins/steelseries/fu-steelseries-hid.c (hunk 4 above)
Comment (gportay) [reply plugins/steelseries/fu-steelseries-hid.c:94]: still debugging.

**Comment by gportay** [comment]
Now working,

```
XPS 13 9370
│
└─Aerox 3 Wireless Mouse via Bluetooth:
      Device ID:          ce2505af510d8b52c393427b0701c9dfbf0260b9
      Current version:    1.8.1
      Vendor:             HIDRAW:0x0111
      GUIDs:              14b18a05-6162-58f6-bae4-a6ff3c8b66c1 ← HIDRAW\VEN_0111&DEV_183A
                          a6afb856-9ef7-53a9-9445-d923fcc6d57d ← HIDRAW\VEN_0111&DEV_183A&REV_00
      Device Flags:       • Updatable
                          • Unsigned Payload
    
```

Next step is to `detach()` to another plugin.

**Comment by gportay** [comment]
Oh, it seems to work! detaching from `FuSteelseriesFizzHid` to `FuSteelseriesFizz`.

But the reciproque is not true; (re)attaching from `FuSteelseriesFizz` to `FuSteelseriesFizzHid` (to post a request telling the user to reconnect the device the old way).

File: plugins/steelseries/steelseries.quirk [hunk 5]
```diff
@@ -41,7 +46,7 @@ Plugin = steelseries
 GType = FuSteelseriesFizz
 Name = Aerox 3 Wireless Mouse
 Icon = input-mouse
-CounterpartGuid = USB\VID_1038&PID_183B
+CounterpartGuid = HIDRAW\VEN_0111&DEV_183A
```
Comment (gportay) [inline plugins/steelseries/steelseries.quirk:49]: However, is is possible to manager multiple `CounterpartGuid` in the quirk file?

```
CounterpartGuid = USB\VID_1038&PID_183B,HIDRAW\VEN_0111&DEV_183A
```
or
```
CounterpartGuid = USB\VID_1038&PID_183B HIDRAW\VEN_0111&DEV_183A
```
or even
```
CounterpartGuid = USB\VID_1038&PID_183B
CounterpartGuid = HIDRAW\VEN_0111&DEV_183A
```

Comment (hughsie) [reply plugins/steelseries/steelseries.quirk:49]: It's probably a comma, but check fu-device.c as I don't think anyone has tried that before.

**Comment by gportay** [comment]
@hughsie: The commit [Allow specifying a list of GUIDs separated by comas](https://github.com/fwupd/fwupd/pull/4698/commits/0f253363243fb30312ce5777cc3bfaa260682fd4) is not necessary; if you think it is useful, I can push it to a dedicated PR and standardize it to every GUIDs quirks.

File: plugins/steelseries/steelseries.quirk (hunk 5 above)
Comment (gportay) [reply plugins/steelseries/steelseries.quirk]: not working; making a commit to support coma separated list of GUIDs.

**Comment by hughsie** [comment]
> if you think it is useful, I can push it to a dedicated PR and standardize it to every GUIDs quirks

Alternate proposal in https://github.com/fwupd/fwupd/pull/4720

**Comment by gportay** [comment]
> Both `fwupd` and `fwupdmgr` claims the USB interface on device open, after the mouse connection is switch from BT to USB; the latter fails with:
> 
> ```
> 13:12:41:0384 FuEngine             failed to add device usb:01:00:01: failed to claim interface 0x03: USB error on device 1038:183a : Resource busy [-6]
> ```

The reverse is also possible.

`fwupd`:

```
14:33:12:0349 FuEngine             failed to add device usb:01:00:01: failed to claim interface 0x03: USB error on device 1038:183a : Resource busy [-6]
```

Only one "process" can claim interface at a time?

Did I make a mistake somewhere?

**Comment by gportay** [comment]
> Only one "process" can claim interface at a time?

However, I am surprised the `mgr` is trying to open the device to `_setup()`. or I may be mistaken.

**Comment by hughsie** [comment]
> However, I am surprised the mgr is trying to open the device

The fwupdmgr doesn't do any device access at all; it's all done in daemon.

**Comment by hughsie** [comment]
> Only one "process" can claim interface at a time?

Have you got a log of the open, close, open, close? it'll fail if you do open, open, close, close.

**Comment by gportay** [comment]
> The fwupdmgr doesn't do any device access at all; it's all done in daemon.

Ok, that makes sense to me.

> Have you got a log of the open, close, open, close? it'll fail if you do open, open, close, close.

`fwupd`: http://ix.io/3ZMy
`fwupdmgr local-install`: http://ix.io/3ZMz

**Comment by gportay** [comment]
The error below is thrown by [fu_usb_device_open()](https://github.com/fwupd/fwupd/blob/wip/gportay/steelseries-add-support-for-hid-devices/libfwupdplugin/fu-usb-device.c#L253) (`fwupdmgr`'s process)
```
16:59:49:0013 FuEngine             failed to add device usb:03:00:01: failed to claim interface 0x03: USB error on device 1038:183a : Resource busy [-6]
```

**Comment by hughsie** [comment]
> The error below is thrown by [fu_usb_device_open()](https://github.com/fwupd/fwupd/blob/wip/gportay/steelseries-add-support-for-hid-devices/libfwupdplugin/fu-usb-device.c#L253) (fwupdmgr's process)

That's `fwupdtool`!

**Comment by gportay** [comment]
> > The error below is thrown by [fu_usb_device_open()](https://github.com/fwupd/fwupd/blob/wip/gportay/steelseries-add-support-for-hid-devices/libfwupdplugin/fu-usb-device.c#L253) (fwupdmgr's process)
> 
> That's `fwupdtool`!

Nah... I copy/paste the wrong line for my hid testing :facepalm:

```
sudo fwupdtool install-blob SteelSeries_Aerox_3_Wireless_Mouse_1.9.3.cab --verbose
```

**Comment by gportay** [comment]
... and it works (again) if using the right utility! Thanks for telling I tested wrong

File: plugins/steelseries/fu-steelseries-fizz-hid.c [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+	if (g_getenv("FWUPD_STEELSERIES_HID_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+	if (!fu_steelseries_fizz_hid_command(device, data, sizeof(data), error))
+		return NULL;
+	if (g_getenv("FWUPD_STEELSERIES_HID_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "Version", data, sizeof(data));
+
+	/* success */
+	return g_strndup((const gchar *)data, sizeof(data));
```
Comment (hughsie) [inline plugins/steelseries/fu-steelseries-fizz-hid.c:88]: fu-common-strsafe perhaps?

**Review by hughsie — APPROVED** [review]

**Review by hughsie — APPROVED** [review]

**Pull request merged**